### DISCUSSION TITLE
Serverless scheduling

### DISCUSSION THREAD

**Discussion by innermatrix** [timgit/pg-boss#530] [Q&A]
My goal is to use the `pg-boss` scheduler in serverless context. The idea is along the lines of 

* Run a lambda on a native AWS timer (using EventBridge)
* Spawn a `pg-boss` instance in that lambda to find all scheduled tasks that need to be run
* Shut down the `pg-boss` instance
* Repeat

But the problem that I have with this is that currently (as I understand from reading the code) the `pg-boss` scheduler will only run a scheduled task if the scheduler invocation takes place <60s after the scheduled task time, and the shortest interval that EventBridge timers can use to invoke lambdas is 1min. So there's a small, but non-zero, probability that a scheduled task will be missed if the timing of the lambda invocation is such that one scheduler check takes place slightly before "scheduled time N + 60s" and the next scheduler check takes place slightly after "scheduled time N+1 + 60s".

Furthermore, maybe I don't want to pay to run a lambda every 1min — maybe I'd rather run the lambda every 10min, as long as I am okay with the implications that has on my scheduled tasks (meaning that I am okay if `5 * * * *` and `6 * * * *` schedules actually both run at ~the same time around 10min after the hour, and a `5,6 * * * *` schedule gets debounced so it only runs once per hour at around 10min after the hour). 

So, what I am wondering is

* Did I understand the behavior of the scheduler correctly wrt only running tasks that are <60s overdue?
* Are there any recommendations for how to run the scheduler in a periodic lambda context (given its restrictions of max rate of one lambda invocation per minute)
* Is there any interest in making changes to the scheduler so that when it's run less frequently than once per minute it operates in a catch-up mode rather than in a skip mode?

Thanks!

**Comment by timgit** [comment]
> Did I understand the behavior of the scheduler correctly wrt only running tasks that are <60s overdue?

Yes, you understood how the scheduler works within a 60s interval.

> Are there any recommendations for how to run the scheduler in a periodic lambda context (given its restrictions of max rate of one lambda invocation per minute)

There is currently no API to do this, but the logic involved could be exposed.

> Is there any interest in making changes to the scheduler so that when it's run less frequently than once per minute it operates in a catch-up mode rather than in a skip mode?

Surely.  Also, if you're interested in submitting a PR, there is also currently interest in supporting per-second cron scheduling, which would be biased in the opposite direction of this.  It does't invalidate what you're doing, but just be aware of that. :)

**Comment by matthiasfeist** [comment]
There's also an issue that might be related to what you want to do: https://github.com/timgit/pg-boss/issues/557
